### DISCUSSION TITLE
Values should be the same data structure

### DISCUSSION THREAD

**Discussion by MartinDawson** [handsontable/hyperformula#690] [General]
### Description

If you do `getCellValue` on an invalid cell then it results in this:

```js
{
  address: "Sheet1!L3"
  message: ""
  type: "REF"
  value: "#REF!"
}
```

Whereas if you perform `getCellValue` on a valid cell it produces a non object like this:

```js
  22
```

Instead of switching data types it's always better to keep consistent and produce the same data type in my opinion so that users don't have to do if statement checks (and also because it's just more developer friendly), in this case I think `getCellValue` should always return an object like so:

```js
{
  value: "#REF!#
}
```

Or if it's valid:

```js
{
  value: 22
}
```
This means that we don't have to do weird checks like `!value.type && ...` or something if we want to check that the cell has computed correctly.

**Comment by wojciechczerniak** [comment]
That depends on your implementation. I don't think that nesting the value will solve a problem here. Since `{ value: * }` may be of any type it will require those additional checks anyway. We will just move the type checking somewhere else.

My take is to use an abstraction over 3rd party lib and have a `getValue` method that will return correctly formatted value that my presentation layer requires. With a single place to make all the checks and format value as I please this is not an issue.

----

Although, it may be a good idea to take a closer look at [`toString()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/toString), [`valueOf()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/valueOf) and [`[Symbol.toPrimitive](hint)`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toPrimitive)

```ts
class DetailedCellError {
  constructor(public value) {}

  toString(): string {
    return this.value
  }

  valueOf(): string {
    return this.value
  }

  [Symbol.toPrimitive](hint): string {
    return this.value
  }
}
```
Result:
```ts
const detailedCellError = new DetailedCellError('#REF!')

console.log(+detailedCellError)        // "#REF!"
console.log(`${detailedCellError}`)    // "#REF!"
console.log(detailedCellError + '')    // "#REF!"
```

Note: We need a feature check for `[Symbol.toPrimitive]` as it's not supported by all the browsers. I think that `toString` and `valueOf` may be enough, `toPrimitive` calls them internally

**Reply by MartinDawson** [reply]
Okay, that makes sense. Thanks!
